compiler/rustc_ty_utils/src/needs_drop.rs

```diff
@@ -16,32 +16,68 @@ fn needs_drop_raw<'tcx>(tcx: TyCtxt<'tcx>, query: ty::ParamEnvAnd<'tcx, Ty<'tcx>
     // If we don't know a type doesn't need drop, for example if it's a type
     // parameter without a `Copy` bound, then we conservatively return that it
     // needs drop.
-    let res = NeedsDropTypes::new(tcx, query.param_env, query.value, adt_fields).next().is_some();
+    let res = NeedsDropTypes::new(
+        tcx,
+        query.param_env,
+        query.value,
+        adt_fields,
+        DropCheckType::NeedsDrop,
+    )
+    .next()
+    .is_some();
     debug!("needs_drop_raw({:?}) = {:?}", query, res);
     res
 }
 
-fn needs_finalizer_raw<'tcx>(tcx: TyCtxt<'tcx>, query: ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool {
-    let adt_fields =
-        move |adt_def: &ty::AdtDef| tcx.adt_finalize_tys(adt_def.did).map(|tys| tys.iter());
-    let res = NeedsDropTypes::new(tcx, query.param_env, query.value, adt_fields).next().is_some();
-    debug!("needs_finalize_raw({:?}) = {:?}", query, res);
-    res
-}
-
 fn has_significant_drop_raw<'tcx>(
     tcx: TyCtxt<'tcx>,
     query: ty::ParamEnvAnd<'tcx, Ty<'tcx>>,
 ) -> bool {
     let significant_drop_fields =
         move |adt_def: &ty::AdtDef| tcx.adt_significant_drop_tys(adt_def.did).map(|tys| tys.iter());
-    let res = NeedsDropTypes::new(tcx, query.param_env, query.value, significant_drop_fields)
-        .next()
-        .is_some();
+    let res = NeedsDropTypes::new(
+        tcx,
+        query.param_env,
+        query.value,
+        significant_drop_fields,
+        DropCheckType::NeedsDrop,
+    )
+    .next()
+    .is_some();
     debug!("has_significant_drop_raw({:?}) = {:?}", query, res);
     res
 }
 
+fn needs_finalizer_raw<'tcx>(tcx: TyCtxt<'tcx>, query: ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool {
+    let finalizer_fields =
+        move |adt_def: &ty::AdtDef| tcx.adt_finalize_tys(adt_def.did).map(|tys| tys.iter());
+    let res = NeedsDropTypes::new(
+        tcx,
+        query.param_env,
+        query.value,
+        finalizer_fields,
+        DropCheckType::NeedsFinalize,
+    )
+    .next()
+    .is_some();
+    debug!("needs_finalize_raw({:?}) = {:?}", query, res);
+    res
+}
+
+enum DropCheckType {
```

Comment:
A maybe-silly question: will this `enum` only ever have two values? If so, is it a boolean in disguise? [The `needs_finalizer` check suggests it is, but I might be over-simplifying!]

Reply:
It is a bool check in disguise yes. That may change in future because newer versions of rustc seem to have introduced the notion of a _significant dropck_: same behaviour as dropck but some drop methods can be marked as _insignificant_ with a `#[insignificant_drop]` attribute.

Reply:
Aha, OK, then let's keep the `enum`!